WalletWasabi/Wallets/BlockDownloadStats.cs

```diff
@@ -0,0 +1,153 @@
+				}
+			}
+
+			var consecutiveFailsPenalty = 0.0;
+			if (ConsecutiveFails > 0)
+			{
+				consecutiveFailsPenalty = (ConsecutiveFails - 1) * Const.BlockDlFailedPenalty *
+				                          Math.Pow(Const.BlockDlFailedMultiplier, ConsecutiveFails - 1);
+			}
+			var result = (int)Math.Round(baseTimeout + consecutiveFailsPenalty);
+			result = result > Const.TimeoutMaxValue ? Const.TimeoutMaxValue : result;
+			var timeoutDeltaLog = lastTimeout >= result ? $"(-{lastTimeout - result})" : $"(+{result - lastTimeout})";
+			Logger.LogDebug($"Timeout: {result}ms {timeoutDeltaLog} / Sample Avg: {Math.Round(AverageLastSample)}ms");
+			return result;
+		}
+	}
+
+	public bool NodeDisconnectStrategy(Node node, int connectedNodes)
+	{
+		lock (Lock)
+		{
+			// Last sample and History for blocks downloaded by this node
+			var nodeDlsSample = Sample.Where(x => x.From.Equals(node.RemoteSocketAddress));
+			var nodeDlsSampleCount = nodeDlsSample.Count();
+			var nodeDlsHistory = History.Where(x => x.From.Equals(node.RemoteSocketAddress));
+			var nodeDlsHistoryCount = nodeDlsHistory.Count();
+			
+			// KeepNodeScore components, the higher the score the higher probability to keep the node
+			var connectedNodesScore = connectedNodes >= Const.ConnectedNodesGoal
+				? (1 - Math.Min(1, (connectedNodes - Const.ConnectedNodesGoal) * Const.TooMuchNodesScoreMultiplier)) * 100 // Low score when too much nodes
+				: (((double)Const.ConnectedNodesGoal - (connectedNodes - 1)) / Const.ConnectedNodesGoal) * 100; // High score when low number of nodes
+			
+			// Low score if high frequency of dl from this node
+			var nodeFrequencyScore = (1 - Math.Min(1, Math.Pow(nodeDlsSampleCount, 2) / Sample.Count())) * 100;
+
+			// Low score if node tends to timeout
+			var nodeHistoryTimeoutsScore = Math.Max(0.15, 1 - Math.Min(1, nodeDlsHistory.Count(x => !x.Success) * Const.NodeHistoryTimeoutsMultiplier / nodeDlsHistoryCount)) * 100;
+			if (ConsecutiveFails > 1)
+			{
+				// Lot of consecutive fails probably means temporary bandwidth/tor issue, avoid disconnecting.
+				nodeHistoryTimeoutsScore = Math.Min(80, nodeHistoryTimeoutsScore * ConsecutiveFails);
+			}
+			
+			var keepNodeScore = Math.Round(Const.NodeFrequencyWeight * nodeFrequencyScore 
+			                               + Const.ConnectedNodesWeight * connectedNodesScore 
+			                               + Const.NodeHistoryTimeoutsWeight * nodeHistoryTimeoutsScore);
+
+			var result = keepNodeScore <= InsecureRandom.Instance.GetInt(0, 101);
+			var blockDlResultLog = Sample.Last().Success ? "Block dl SUCCESS" : "Block dl FAILED";
+			Logger.LogDebug(result
+				? $"{blockDlResultLog} - Node: Disconnect - Score: {keepNodeScore} - Nb Nodes remaining: {connectedNodes - 1}"
+				: $"{blockDlResultLog} - Node: Keep - Score: {keepNodeScore} - Nb Nodes: {connectedNodes}");
+			return result;
+		}
+	}
+	
+	private class BlockDl
```

Comment:
Use class names that are much descriptive. Try to avoid abbreviations, again the reason is readability. 
For example: `NodeBlockDownloadResult`